sdk/eventgrid/azure-messaging-eventgrid/src/main/java/com/azure/messaging/eventgrid/implementation/CloudEventTracingPipelinePolicy.java

```diff
@@ -0,0 +1,76 @@
+import com.azure.core.util.tracing.TracerProxy;
+import reactor.core.publisher.Mono;
+
+import java.nio.charset.StandardCharsets;
+
+import com.azure.messaging.eventgrid.CloudEvent;
+/**
+ * This pipeline policy should be added after OpenTelemetryPolicy in the http pipeline.
+ *
+ * It checks whether the {@link HttpRequest} headers have "traceparent" or "tracestate" and whether the serialized
+ * http body json string for a list of {@link CloudEvent} instances has place holders
+ * {@link Constants#TRACE_PARENT_PLACEHOLDER} or {@link Constants#TRACE_STATE_PLACEHOLDER}.
+ * The place holders will be replaced by the value from headers if the headers have "traceparent" or "tracestate",
+ * or be removed if the headers don't have.
+ *
+ * The place holders won't exist in the json string if the {@link TracerProxy#isTracingEnabled()} returns false.
+ */
+public class CloudEventTracingPipelinePolicy implements HttpPipelinePolicy {
+    @Override
+    public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineNextPolicy next) {
+        final HttpRequest request = context.getHttpRequest();
+        final HttpHeader contentType = request.getHeaders().get(Constants.CONTENT_TYPE);
+        StringBuilder bodyStringBuilder = new StringBuilder();
+        if (TracerProxy.isTracingEnabled() && contentType != null &&
+            Constants.CLOUD_EVENT_CONTENT_TYPE.equals(contentType.getValue())) {
+            return request.getBody().map(byteBuffer -> bodyStringBuilder.append(new String(byteBuffer.array(),
+                StandardCharsets.UTF_8)))
+                .then(Mono.fromCallable(() -> replaceTracingPlaceHolder(request, bodyStringBuilder)))
+                .then(next.process());
+        }
+        else {
+            return next.process();
+        }
+    }
+
+    /**
+     *
+     * @param request The {@link HttpRequest}, whose body will be mutated by replacing traceparent and tracestate
+     *                placeholders.
+     * @param bodyStringBuilder The {@link StringBuilder} that contains the full HttpRequest body string.
+     * @return The new body string with the place holders replaced (if header has tracing)
+     * or removed (if header no tracing).
+     */
+    static String replaceTracingPlaceHolder(HttpRequest request, StringBuilder bodyStringBuilder) {
+        final int traceParentPlaceHolderIndex = bodyStringBuilder.indexOf(Constants.TRACE_PARENT_PLACEHOLDER);
+        if (traceParentPlaceHolderIndex >= 0) { // There is "traceparent" placeholder in body, replace it.
+            final HttpHeader traceparentHeader = request.getHeaders().get(Constants.TRACE_PARENT);
+            bodyStringBuilder.replace(traceParentPlaceHolderIndex,
+                Constants.TRACE_PARENT_PLACEHOLDER.length() + traceParentPlaceHolderIndex,
+                traceparentHeader != null
+                    ? String.format(",\"%s\":\"%s\"", Constants.TRACE_PARENT, traceparentHeader.getValue())
+                    : "");
```

Comment:
Do we want the request headers to have the keys of `tracestate` and `traceparent` with empty values ?
 >traceparentHeader != null
                    ? String.format(",\"%s\":\"%s\"", Constants.TRACE_PARENT, traceparentHeader.getValue())
                    : "")

Reply:
Clarified offline, this is not updating the request headers.